README.md
# mcrconTLS
Python implementation of [Minecraft RCON](https://wiki.vg/RCON) with TLS.

Inspired by the work of [barneygale](https://github.com/barneygale/MCRcon) and [Tiiffi](https://github.com/Tiiffi/mcrcon), I decided to try my hand at my own implementation.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2017-05-12
### Added
- This changelog
- send and receive methods for the MCRCONTLSServer class
### Changed
- Moved connect() implememtation to MCRCONTLSClient
    - How the connection is initiatied should be the responsibility of the client. This change removes functionality from the shared McRconTLS class that can't be used by the server
- MCRCONTLSServer can now handle connections from multiple clients




client/mcrconTLS.py

# Implementation of Minecraft RCON supporting TLS
# Based on barneygale's implementation: https://github.com/barneygale/MCRcon/blob/master/mcrcon.py

import socket
import ssl
import struct
from collections import namedtuple


# https://docs.python.org/3/library/argparse.html
# https://docs.python.org/3/library/asyncio.html
# https://docs.python.org/3/library/ssl.html
# https://docs.python.org/3/library/struct.html
# https://docs.python.org/3/library/socket.html


class IncompletePacketException(Exception):
    '''RCON packet is not fully formed'''
    def __init__(self, m):
        self.minimum = m
    pass

class REQUEST_ID:
    MESSAGE     = 0
    END_MESSAGE = 1

class Packet:
    '''
    A symple RCON packet
    Specification: https://wiki.vg/RCON
    '''
    class TYPE:
        LOGIN   = 3
        COMMAND = 2
        MULTI   = 0

    def __init__(self, id, ptype, payload=b''):
        self.id = id
        self.ptype = ptype
        self.payload = payload

class McRconTLS:
    '''Communicate via RCON over TLS'''
    def __init__(self, host, port, password=''):
        
        # if mode == 0:
        #     self.init_client()
        # else:
        #     self.init_server()

        self.host = host
        self.port = port
        self.password = password

    
    def console(self):
        '''Simple prompt for user input'''
        return input('> ')
    
    # construct a packet as a struct and return it
    def packet_encode(self, packet):
        '''Encode an RCON packet'''

        data = struct.pack('<ii', packet.id, packet.ptype) + packet.payload + b'\x00\x00'
        return struct.pack('<i', len(data)) + data
    
    def packet_decode(self, data):
        '''Decode an RCON packet'''

        if len(data) < 14:
            # minimum packet size (empty payload) is 14 bytes
            raise IncompletePacketException(14)

        pklen = struct.unpack('<i', data[0:4])[0]+4
        #print(f'pklen={pklen}')

        if len(data) < pklen:
            # received less data than the packet header states
            raise IncompletePacketException(pklen)
        
        pkid, pktype = struct.unpack('<ii', data[4:12])
        payload = data[12:pklen-2]
        padding = data[pklen-2:]
        
        #print(f'Packet: {pklen}, {pkid}, {pktype}, {payload}, {padding}')
        
        assert padding == b'\x00\x00'

        return Packet(pkid, pktype, payload)
    
    def packet_recv(self, sock):
        '''Receive packets from the given socket. Returns a Packet object'''

        data  = b''

        while True:
            try:
                return self.packet_decode(data), len(data)
            except IncompletePacketException as e:
                while len(data) < e.minimum:
                    data += sock.recv(e.minimum - len(data))
        # data = sock.recv(4096)
        # print(f'data={data}')

        # return self.decode(data)

    def packet_send(self, sock, packet):
        '''Send the packet over the given socket.'''

        enc = self.packet_encode(packet)
        sock.sendall(enc)


client/McRconTLSClient.py

# McRconTLS client

import ssl, socket
import threading
from mcrconTLS import *

class MCRCONTLSClient:

    host = None
    port = None
    client = None


    def __init__(self, host='localhost', port='25575'):
        self.host = host
        self.port = port

    def login(self, password=''):
        self.client = McRconTLS(self.host, self.port, password='')

        return True

    def command_send(self, ssock, cmd):
        self.client.packet_send(
            ssock,
            Packet(REQUEST_ID.MESSAGE, Packet.TYPE.COMMAND, cmd.encode())
        )
        # send message-terminating packet
        self.client.packet_send(
            ssock,
            Packet(REQUEST_ID.END_MESSAGE,Packet.TYPE.COMMAND)
        )

    def connect(self,verify_mode):
        '''Attempt to establish a secured connection to the server'''

        context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH, cafile='ssl/serverCert.crt')
        context.load_verify_locations('./ssl/serverCert.crt')
        context.check_hostname = False
        
        if verify_mode == 0:
            context.verify_mode = ssl.CERT_NONE
        elif verify_mode == 1:
            context.verify_mode = ssl.CERT_OPTIONAL
        elif verify_mode  == 2:
            context.verify_mode = ssl.CERT_REQUIRED


        try:
            ssock = context.wrap_socket(socket.create_connection((self.host, self.port)))

            # Enter login credentials
            if (ssock):
                if self.login():
                    print(f'logged into server. {self.host}:{self.port}')
                else:
                    return

                    # start a thread to receive output from the server
                    threading.Thread(target=self.receive_server_output, args=(ssock,), daemon=True)

            # with socket.create_connection((self.host, self.port)) as sock:
            #     with context.wrap_socket(sock) as ssock:
            while (ssock):
                if (command := input('> ')) != 'exit':
                    # send the command packet
                    self.command_send(ssock, command)

        except ConnectionResetError:
            print('Server closed the connection')
        except ConnectionAbortedError:
            print('Connection was aborted by the server')
        except ConnectionRefusedError:
            print('Connection was refused by the server.')

    def receive_server_output(self, ssock):
        while ssock:
            data, size = self.client.packet_recv()
            if size > 14:
                print(f'{data}\n')


if __name__ == '__main__':
    client = MCRCONTLSClient('192.168.56.1', 25576)

    client.connect(verify_mode=0)


server/McRconTlsServer.py
# mcrconTLS.py

import subprocess
import threading
import socket
import ssl
import queue
import re
import sys

from mcrconTLS import *

class MCRCONTLSServer:
    '''A wrapper for minecraft server that supports RCON over TLS.'''

    connections = []
    command_queue = queue.Queue()
    command_queue_condition = threading.Condition()

    RCON_PORT = 25576 # default port number is 25575. Using 25576 for encrypted comms

    CERT = './ssl/serverCert.crt'
    KEY  = './ssl/serverKey.key'

    def __init__(self, cert=None, key=None):
        print("Initialized mcrconTLS")

        if cert:
            self.CERT = cert
        
        if key:
            self.KEY = key
    
    # def send_response(self):
    #     '''Send server response and send it to the client'''

    def read_stdout(self):
        while True:
            out = self.minecraft.stdout.readline()

            if out:
                print(f'[stdout]' + out, end='')

    def rcon_queueWriter(self):
        while self.minecraft.poll() == None:

            with self.command_queue_condition:

                while self.command_queue.empty():
                    self.command_queue_condition.wait()

                cmd = self.command_queue.get()
                #print(f'[{__name__}] writing {cmd} to mcstdin -> {cmd.encode()}')
                
                print(cmd, file=self.minecraft.stdin)
                #self.minecraft.stdin.write(cmd.encode())
                self.minecraft.stdin.flush()
                
    def rcon_server(self):
        print('Opening TLS encrypted channel')
        # https://docs.python.org/3/library/ssl.html
        context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
        context.verify_mode = ssl.CERT_NONE
        context.load_cert_chain(certfile='./ssl/serverCert.crt', keyfile='./ssl/serverKey.key') # REPLACE WITH VALID CERT SIGNED BY CA

        IP = socket.gethostbyname(socket.gethostname())

        ssock = context.wrap_socket(
            socket.create_server((IP, self.RCON_PORT)),
            server_side=True
        )

        if (ssock):
            # IP = socket.gethostbyname(socket.gethostname())
            # sock.bind((IP, self.RCON_PORT))
            print(f'Listening for RCON connections on {IP}:{self.RCON_PORT}')
            # sock.listen()
            while True:
        #     with socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0) as sock:
        #         IP = socket.gethostbyname(socket.gethostname())
        #         sock.bind((IP, self.RCON_PORT))
        #         print(f'Listening for RCON connections on {IP}:{self.RCON_PORT}')
        #         sock.listen()

        #         with context.wrap_socket(sock, server_side=True) as ssock:
                conn, addr = ssock.accept()
                print (f'TLS connection from {addr}')

                rcon = McRconTLS(addr[0], addr[1])

                self.connections.append((rcon, conn))

                # start a thread with a new rcon session.
                threading.Thread(target=self.rcon_session, args=(rcon, conn), daemon=True).start()
    
    # begin a session with the given rcon object on the specified connection
    def rcon_session(self, rcon, conn):
        '''Start threads for sending to and receiving from client'''
        threading.Thread(target=self.rcon_session_recv, args=(rcon, conn)).start()
        threading.Thread(target=self.rcon_session_send, args=(rcon, conn), daemon=True).start()

    def rcon_session_recv(self, rcon, conn):
        '''Routine to recieve incoming messages from a client'''
        try:
            while conn and self.minecraft.poll() == None:
                p, size = rcon.packet_recv(conn)
                #print(f'{p.id}, {p.ptype}, {p.payload.decode()}')
                if size > 14:
                    with self.command_queue_condition:
                        self.command_queue.put(p.payload.decode())
                        #print(f'[{__name__}] put {p.payload} in queue')

                        self.command_queue_condition.notifyAll()
            print(f'RCON Connection to {rcon.host}:{rcon.port} closed.')
        except ConnectionResetError:
            print(f'[{rcon.host}:{rcon.port}] connection closed by client.')


    def rcon_session_send(self, rcon, conn):
        '''Routine to send the server stdout to the client'''
        try:
            while conn and self.minecraft.poll() == None:
                rcon.packet_send(
                    conn,
                    Packet(REQUEST_ID.MESSAGE, Packet.TYPE.MULTI, self.minecraft.stdout.readline().encode())
                )
        except ConnectionResetError:
            print(f'[{rcon.host}:{rcon.port}] connection closed by client.')

    def start(self):
        self.start_minecraft('java -jar ./minecraft_server.jar nogui')

  
    def start_minecraft(self, command):
        '''Start the Minecraft server a subprocess, then start rcon threads'''
        print(command)

        self.minecraft = subprocess.Popen(command,
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.STDOUT,
                                    stdin=subprocess.PIPE,
                                    text=True)

        self.start_rcon_server()

        threading.Thread(target=self.read_stdout, daemon=True).start()

    def start_rcon_server(self):
            '''Starts the main RCON server thread, as well as the thread that writes to the RCON command queue'''
            threading.Thread(target=self.rcon_server, name="RCONServer").start()
            threading.Thread(target=self.rcon_queueWriter, name='rcon_queueWriter', daemon=True).start()


if __name__ == "__main__":
    #print(mcrconTLS.__doc__)

    server = MCRCONTLSServer()
    server.start()

    